File: app/src/main/java/com/nuhkoca/myapplication/paging/VideoResultDataSourceFactory.java
package com.nuhkoca.myapplication.paging;


import com.nuhkoca.myapplication.data.remote.video.VideoResponse;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.paging.DataSource;

/**
 * A factory class for {@link PageKeyedVideosDataSource}
 *
 * @author nuhkoca
 */
public class VideoResultDataSourceFactory extends DataSource.Factory<Integer, VideoResponse> {

    private MutableLiveData<PageKeyedVideosDataSource> mPageKeyedVideosDataSourceMutableLiveData;
    private PageKeyedVideosDataSource pageKeyedVideosDataSource;

    /**
     * A default constructor that injects required dependencies
     *
     * @param pageKeyedVideosDataSource represents an instance {@link PageKeyedVideosDataSource}
     */
    @Inject
    VideoResultDataSourceFactory(PageKeyedVideosDataSource pageKeyedVideosDataSource) {
        this.mPageKeyedVideosDataSourceMutableLiveData = new MutableLiveData<>();
        this.pageKeyedVideosDataSource = pageKeyedVideosDataSource;
    }

    /**
     * Applies the given function to each value emitted by DataSources produced by this Factory.
     *
     * @return a {@link DataSource}
     */
    @NonNull
    @Override
    public DataSource<Integer, VideoResponse> create() {
        mPageKeyedVideosDataSourceMutableLiveData.postValue(pageKeyedVideosDataSource);
        return pageKeyedVideosDataSource;
    }

    /**
     * Returns {@link VideoResultDataSourceFactory#mPageKeyedVideosDataSourceMutableLiveData}
     *
     * @return {@link VideoResultDataSourceFactory#mPageKeyedVideosDataSourceMutableLiveData}
     */
    public MutableLiveData<PageKeyedVideosDataSource> getPageKeyedVideosDataSourceMutableLiveData() {
        return mPageKeyedVideosDataSourceMutableLiveData;
    }

    /**
     * Returns {@link VideoResultDataSourceFactory#pageKeyedVideosDataSource}
     *
     * @return {@link VideoResultDataSourceFactory#pageKeyedVideosDataSource}
     */
    public PageKeyedVideosDataSource getPageKeyedVideosDataSource() {
        return pageKeyedVideosDataSource;
    }
}

File: app/src/main/java/com/nuhkoca/myapplication/paging/VideoResultDataSourceFactory.kt
package com.nuhkoca.myapplication.paging

import androidx.lifecycle.MutableLiveData
import androidx.paging.DataSource
import com.nuhkoca.myapplication.data.remote.video.VideoResponse
import javax.inject.Inject

/**
 * A factory class for [PageKeyedVideosDataSource]
 *
 * @author nuhkoca
 */
class VideoResultDataSourceFactory @Inject constructor(
    private val pageKeyedVideosDataSource: PageKeyedVideosDataSource
) : DataSource.Factory<Int, VideoResponse>() {

    /**
     * Returns [VideoResultDataSourceFactory.mPageKeyedVideosDataSourceMutableLiveData]
     *
     * @return [VideoResultDataSourceFactory.mPageKeyedVideosDataSourceMutableLiveData]
     */
    val pageKeyedVideosDataSourceMutableLiveData: MutableLiveData<PageKeyedVideosDataSource> = MutableLiveData()

    /**
     * Applies the given function to each value emitted by DataSources produced by this Factory.
     *
     * @return a [DataSource]
     */
    override fun create(): DataSource<Int, VideoResponse> {
        pageKeyedVideosDataSourceMutableLiveData.postValue(pageKeyedVideosDataSource)
        return pageKeyedVideosDataSource
    }
}
